src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,258 @@
+                let layout = self.ecx.layout_of(ty.builtin_deref(true).unwrap().ty)?;
+                if layout.is_unsized() {
+                    if let ty::Dynamic(..) = self.ecx.tcx.struct_tail(layout.ty).sty {
+                        if let Ok(vtable) = meta.unwrap().to_ptr() {
+                            // explitly choose `Immutable` here, since vtables are immutable, even
+                            // if the reference of the fat pointer is mutable
+                            self.intern(vtable, Mutability::Immutable)?;
+                        }
+                    }
+                }
+            }
+            let mplace = self.ecx.ref_to_mplace(value)?;
+            // Check if we have encountered this pointer+layout combination before.
+            // Only recurse for allocation-backed pointers.
+            if let Scalar::Ptr(ptr) = mplace.ptr {
+                let mutability = match mutability {
+                    hir::Mutability::MutImmutable => Mutability::Immutable,
+                    hir::Mutability::MutMutable => Mutability::Mutable,
+                };
+                match self.intern(ptr, mutability)? {
+                    // No need to recurse, these are interned already and statics may have
+                    // cycles, so we don't want to recurse there
+                    Some(IsStaticOrFn) => {},
+                    // intern everything referenced by this value. The mutability is taken from the
+                    // reference. It is checked above that mutable references only happen in
+                    // `static mut`
+                    None => self.ref_tracking.track((mplace, mutability), || ()),
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+pub fn intern_const_alloc_recursive(
+    ecx: &mut CompileTimeEvalContext<'a, 'mir, 'tcx>,
+    def_id: DefId,
+    ret: MPlaceTy<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+) -> EvalResult<'tcx> {
+    let tcx = ecx.tcx;
+    // start out immutable and only let `UnsafeCell` change that
+    let is_freeze = ret.layout.ty.is_freeze(tcx.tcx, param_env, tcx.span);
+    let (mutability, intern_mode) = match (tcx.static_mutability(def_id), is_freeze) {
+        (Some(hir::Mutability::MutImmutable), true) |
+        (None, true) => (Mutability::Immutable, InternMode::StaticOrConst),
+        // constants and statics are only mutable if they have interior mutability. Even then, only
+        // the outermost allocation needs to be mutable, since everything else is guaranteed behind
+        // an immutable reference and has no interior mutability (checked by const_qualif).
+        (Some(hir::Mutability::MutImmutable), false) |
+        (None, false) => (Mutability::Mutable, InternMode::StaticOrConst),
+        // `static mut` doesn't care about interior mutability, it's mutable anyway
+        (Some(hir::Mutability::MutMutable), _) => (Mutability::Mutable, InternMode::StaticMut),
+    };
```

Comment:
What does `static_mutability` return for a `const`? Seems like it is `None`? Please say that explicitly.

Also WTF, how are constants with interior mutability a thing? That's so weird.^^

Reply:
> Also WTF, how are constants with interior mutability a thing? That's so weird.^^

Ah right, we just copy the bytes computed by the const initializer to where it is used and then they can be mutated.

But it is by a mere accident that we can treat `static` and `const` the same, namely that `&Cell::new(0)` is not allowed in a `static` for no good reason. Conceptually I think it makes more sense (in `InternMode`) to distinguish just `const` from all `static`/`static mut`; the former must be immutable internally but the latter can have `UnsafeCell` (or even start out mutable).  Our current implementation indeed justifies re-using the `const` check for `static`, but I find that makes it much harder to understand why we are even doing these checks.

Generally your justification for the checks here often is "because const qualif"; I think it'd be better if you could justify them "because here's why that's unsound" -- after all, the goal is to make const qualif say "we need to do this because otherwise interning ICEs [for good reasons]", it doesn't help if the two just point at each other. ;)

Reply:
`&Cell::new(0)` is forbidden in `static`s because of the lack of `Sync`, but yea, generally interior mutability in statics should be ok.

I agree that we should have an independent notion of soundness here. It's just very hard to test that we are doing the right thing.